app/mailers/producer_mailer.rb

```diff
@@ -0,0 +1,37 @@
+
+class ProducerMailer < Spree::BaseMailer
+
+  def order_cycle_report(producer, order_cycle)
+    @producer = producer
+    @coordinator = order_cycle.coordinator
+    @order_cycle = order_cycle
+
+    subject = "[#{Spree::Config[:site_name]}] Order cycle report"
+
+    # if @order_cycle.distributors.any?
+    #   first_producer = @order_cycle.distributors.first
+    #   @distribution_date = @order_cycle.pickup_time_for first_producer
+    #   subject += " for #{@distribution_date}" if @distribution_date.present?
+    # end
+
+    # TODO: what if producer handling orders into multiple order cycles?
+    @orders = Spree::Order.complete.not_state(:canceled).managed_by(@producer.owner)
```

Comment:
It looks like this query will fetch more orders than we want:

- If the owner of the producer owns more producers, then we'll get orders involving those producers, too.
- Orders from any order cycle will be returned, not just the current order cycle.

What we want to fetch (as I understand it) are a list of complete orders from this order cycle, and then pull out all the line items that come from the specified producer. A first attempt at this might look like this:

```ruby
orders = Spree::Order.where(order_cycle: order_cycle).complete
@line_items = orders.map(&:line_items).flatten.select { |o| o.variant.product.supplier == producer }
```

However, fetching line items this way is hugely inefficient, because `map` will trigger a separate SQL query for each order we're looking at (loading the line items), and then `select` will trigger *three* queries for each order, to fetch the variant, then the product, then the supplier. All we care about is getting the line items we want, so we can perform this query purely in SQL and avoid going back and forth between Ruby and SQL.

To do this, I like to start by thinking in terms of the resource I want to get back at the end, ie. line items. If we were to phrase that in terms of the line items we want, I'd say: Line items within the order cycle we're looking at, and with the producer of the product being the producer we're looking at. We can construct this query in SQL like this:

```ruby
Spree::LineItem.
    joins(:order => :order_cycle, :variant => :product).
    where('order_cycles.id = ?', order_cycle).
    where('spree_products.supplier_id = ?', producer)
```

I think this should do the trick, but I haven't tried it myself. Have a play and let me know how you go.